### ISSUE TITLE
-Wunsequenced is too strict on pre-increments

### ISSUE THREAD

**Issue by Endilll** [llvm/llvm-project#63935]
Consider the following example that uses only pre-increments:
```cpp
constexpr int inc(int i)
{
  auto f = [](int i, int j){
    return i + j;
  };
  return f(++i, ++i);
}

int main() {
    constexpr int h = 3;
    static_assert(inc(h) == 9, "");
    int i = 3;
    return inc(i);
}
```
We issue the following diagnostic since C++17:
```cpp
<source>:6:12: warning: multiple unsequenced modifications to 'i' [-Wunsequenced]
  return f(++i, ++i);
           ^    ~~
```
Despite the fact that [P0145R3](https://wg21.link/p0145r3) "Refining Expression Evaluation Order for Idiomatic C++" and [P0400R0](https://wg21.link/p0400r0) "Wording for Order of Evaluation of Function Arguments" were adopted for C++17. Since [then](https://timsong-cpp.github.io/cppwp/n4659/expr.call#5) [[expr.call]](http://eel.is/c++draft/expr.call#7.sentence-2) states the following: `The initialization of a parameter, including every associated value computation and side effect, is indeterminately sequenced with respect to that of any other parameter`. Note that "indeterminately sequenced" means that `f` is called with `(4, 5)` or `(5, 4)`, and `i == 5` by the time we return from `inc()`, leading to static_assert not firing, and program returning 9.

Compilers agree, except for GCC runtime evaluation, which produces 10: https://godbolt.org/z/M8r8rvndT. I don't see an interpretation of the Standard since 17 that makes 10 a conforming result, so I believe there is a bug to file against GCC.

That said, I consider this code bug-prone. But "unsequenced" part of diagnostic message doesn't hold since C++17, and can even be misleading, because it shouldn't take users too many hops to reach a conclusion that unsequenced modification means a data race.

**Comment by dwblaikie** [comment]
Yeah, I think you're right all there. Diagnostic was probably implemented pre-17, and needs a tweak when running in 17+ mode.

**Comment by poyaoc97** [comment]
Will upload a fix once I finish the release note.

**Comment by Endilll** [comment]
I've discussed this with Aaron. Since unsequenced access is still a thing, what we need is a new diagnostic for indeterminately sequenced accesses. Apparently, logic begind `-Wunsequenced` is old enough to lack the notion of "indeterminate sequencing", which makes fixing this non-trivial.

**Comment by poyaoc97** [comment]
A potential fix: https://reviews.llvm.org/D156057
